### PR TITLE
Updating install instructions and splitting config files

### PR DESCRIPTION
**Pull request by betatim**
The instructions in the `README` were very out of date. Updated them which also greatly simplifies them.

Did a bit more splitting of config files to make it easier to compose them. Can someone with a linux host test if the `local_config.py` works as expected? I also changed the default to not use a whitelist or the whitelist reloader. This is probably an "expert" option, so there is a config file to add in to enable it if you know what you are doing.

### CODE REVIEW

**Comment by betatim** [comment]
Maybe we can merge this before #71 or I add the sentence about taking care of updating `dockerspawner` to this PR.

**Comment by betatim** [comment]
Review and merge please.

File: README.md [hunk 1]
```diff
@@ -59,30 +51,41 @@ In order to deploy your own `everware` instance, you have to:
...
 ## Development
 
-To run our tests you will need [selenium] and [phantomjs]. On OS X you can install them with:
+Follow the instructions for deploying your own everware instance. In
+addition to run our tests you will need [selenium] and [firefox]. On
+OS X you can install selenium with:
 
 ```
 brew install selenium-server-standalone
-brew install phantomjs
```
Comment (astiunov) [inline README.md:78]: Why did you remove it? It's much more convenient to use it instead of firefox as for me

**Comment by astiunov** [comment]
I suggest to get rid of `source env.sh` and make an analogue of it in github's config

File: etc/base_config.py [hunk 2]
```diff
@@ -1,11 +1,10 @@
-import os
-import everware
+# Basic configuration, you should not use this directly
+# instead checkout local_config.py or local_dockermacine_config.py
+
 import jupyterhub.handlers.pages
 import jupyterhub.handlers.base
```
Comment (astiunov) [inline etc/base_config.py]: Don't need this anymore

File: README.md (hunk 1 above)
Comment (betatim) [inline README.md:78]: I saw that the [tests use firefox](https://github.com/betatim/everware/blob/install-instructions/frontend_tests/test_generator.py#L13-L16) when not running on travis, so we don't need it anymore. Think I agree that phantomjs is nicer but that should be a different PR.

File: etc/base_config.py (hunk 2 above)
Comment (betatim) [inline etc/base_config.py]: good catch!

**Comment by betatim** [comment]
What do you mean get rid of `source env.sh`? We need some way to set environment variables with the secret information without having the information in the terminal's history. I think I prefer having one file `env.sh` that sets all those secret related variables, for example you can use it to set the jupyerhub `COOKIE_SECRET` or API token in addition to the :octocat: secrets

**Comment by astiunov** [comment]
I mean that we have to to `source env.sh` before everware starts. Maybe it's better to move executing of `env.sh` to github config?

**Comment by anaderi** [comment]
@astiunov do you mean jupyterhub by `github` config?

anyway having a .sh file is helpful for me to play with it using bash, for example, I have DOCKER_HOST that point to swarm in the env.sh and when I want to connect to docker swarm I simply `source env.sh` and then run `docker something`

**Comment by astiunov** [comment]
yeah, I mean `etc/github_auth.py`

**Comment by anaderi** [comment]
maybe moving `GITHUB_*` params to github-specific config is not that bad idea, but then we end up with one version of file that is generic (like .orig or .tpl) and another that user should edit for his specific instance, and shouldn't commit no matter what.
I'd vote to have as less as possible such files. and env.sh seems to be a good candidate for that.

**Comment by astiunov** [comment]
No, I don't vote for removing env.sh :) I talk about how we run everware: now I have to to something like
`source env.sh && everware-server ...`. I suggest to move this `source env.sh` to `github_auth.py` by using `subprocess.call(["source", "env.sh"])` for example

**Comment by betatim** [comment]
I prefer the explicit `source env.sh` step.

**Comment by anaderi** [comment]
`subprocess.call` would create subprocess and all changes of environment will die after it stops )
why do you `source env.sh && everware-server ...` instead of `make run`?

**Comment by betatim** [comment]
Also, I would vote to move the discussion on `source env.sh` to a new issue as it is somewhat orthogonal to getting the docs up to date with the current state of the repo.

**Comment by astiunov** [comment]
But in README.md there is no info about `make run`. It also doesn't mention that we have to `source` every time we launch everware

**Comment by anaderi** [comment]
actually a good point, why don't tell people about `make install` and `make run`?

**Comment by betatim** [comment]
Using `make install` now but I prefer giving the explicit command for running `everware-server`. It let's you see the log output directly on the console and leaves it to a human to decide if they use a VM to run docker or not.

Also moved the instructions about `env.sh` so that it should be clearer that you need to run this every time.

**Comment by betatim** [comment]
The last commit removes the left over `jupyterhub_config.py` and edits the `Makefile`.

However this means that `make run` is now broken. Does someone know enough `Makefile` language to make a robust `if` so that we can detect the OS and depending on that choose the config file that is used? I don't think checking for `docker-machine` is good enough as there are linux users who have `docker-machine` (for EC2 etc) but locally don't need it. Is looking at `uname` a better test?

File: etc/base_config.py [hunk 3]
```diff
@@ -1,10 +1,8 @@
-import os
-import everware
-import jupyterhub.handlers.pages
-import jupyterhub.handlers.base
+# Basic configuration, you should not use this directly
+# instead checkout local_config.py or local_dockermacine_config.py
 
+import jupyterhub.handlers.pages
```
Comment (astiunov) [inline etc/base_config.py]: Did you forget to remove it again?)

File: Makefile [hunk 4]
```diff
@@ -55,7 +55,6 @@ install:  ## install everware
 	${PYTHON} setup.py js
 
 	if [ ! -f env.sh ] ; then cp env.sh.orig env.sh ; fi
-	if [ ! -f jupyterhub_config.py ] ; then cp jupyterhub_config.py.orig jupyterhub_config.py ; fi
 	if [ ! -f whitelist.txt ] ; then cp whitelist.txt.orig whitelist.txt ; fi
```
Comment (astiunov) [inline Makefile]: I suppose we should delete it too

File: README.md [hunk 5]
```diff
@@ -28,61 +28,64 @@ original everware team consists of
...
- - [Create a Github OAuth application](https://github.com/settings/applications/new) with URL `http://localhost:8000/` and callback URL `http://localhost:8000/hub/oauth_callback`
- - Enter you OAuth information into `env.sh` and `source env.sh`
- - Install everware
+ - [Create a Github OAuth application](https://github.com/settings/applications/new)
+   with URL `http://localhost:8000/` and callback URL
+   `http://localhost:8000/hub/oauth_callback`
+ - Enter you OAuth information into `env.sh` and `source env.sh`. You will
+   have to set these environment variables every time before running `everware`
+
+ - If you are on **Mac OS** make sure the VM in which docker runs is
```
Comment (astiunov) [inline README.md:59]: Should we add something about Windows?

**Comment by astiunov** [comment]
Actually we can implement such test with any language (python?) and just exec it from Makefile

File: etc/base_config.py (hunk 3 above)
Comment (betatim) [inline etc/base_config.py]: Yes, coding with my eyes shut apparently ... :(

File: README.md (hunk 5 above)
Comment (betatim) [inline README.md:59]: As I have no knowledge at all about windows and also can't test it I created an issue instead: #92

**Comment by betatim** [comment]
For the moment I prefer explicitly typing the command instead of using `make run`. (Using `make` to run services just feels weird to me.) I created an issue to track progress on making `make run` smart enough to detect which one of the config files it should use: #94

**Comment by betatim** [comment]
Temporarily removing the two make targets which won't work until #94 is implemented. I copied the code for them into the issue so people can pick it up when they add it back.

**Comment by anaderi** [comment]
I've added OS detection, in #97. 
so you can specify which config file to use in `uname -s` if/else block

I'm not sure I follow your logic of removing the targets? 
just a brief motivation I have: a) when debugging everware I'm used to start/stop it rather frequently, so typing full line of arguments is somewhat boring, b) list of arguments might change (e.g. --no-ssl option came out of blue one day, and updating README every time jupyterhub changes is pointless and c) there is no special need to use any kind of daemon/rc scripts to run everare as a daemon -- all of them are more/less the same as Makefile. At least the latter is true until we decide to have one, so I'd make PR with such file and then remove part of Makefile that I extensively use in production. Not vice versa.

**Comment by betatim** [comment]
The motivation for removing them was that at the time they didn't work and I think having stuff that is broken by default isn't good. After #97 the targets should work again.

**Comment by astiunov** [comment]
#97 was merged, would you rebase this pr?

**Comment by betatim** [comment]
Travis is happy now. There were a few failures that were fixed by restarting the build. So I think these are due to timeouts occurring. @anaderi could you point to a few places where there are limits that we can experiment with increasing?

**Comment by betatim** [comment]
Otherwise this can be merged I think.

**Comment by anaderi** [comment]
@betatim , I guess you also need to make some adjustment in order to fix `make run`, since it starts

```
${EXECUTOR} ${OPTIONS}
```

and 

```
OPTIONS = --debug --port 8000 --no-ssl --JupyterHub.hub_ip=${IP}
```

**Comment by betatim** [comment]
Split `make run` into `make run-linux` and `make run-dockermachine`.

**Comment by anaderi** [comment]
there is one more target that needs fixing:

```
run-daemon: clean
    source ./env.sh && \
        ${EXECUTOR} ${OPTIONS} >> ${LOG}  2>&1 &
```

and I'm also worried about, say, SSL-enabled configurations? I guess adding targets like 'run-linux-ssl', 'run-docker-machine-ssl', 'run-daemon-linux', etc is a bit weird. 
Maybe we leave just 2 targets (`run` and `run-daemon`) and allow users to pass name of configuration file as an argument to `Makefile` i.e.
`make run -e CONFIG=etc/local_dockermachine_config.py`)?

**Comment by betatim** [comment]
Having the default/recommended ways to run `everware-server` include the "just turn off SSL" is probably a bad idea :worried: 

Modified the run-daemon target. It assumes that you are on linux, and have configured SSL.

I am -0 on `make run -e CONFIG=etc/local_dockermachine_config.py` it is as inconvenient to type as `everware-server -f etc/local_config.py` and adds a level of indirection.

Another option could be to provide `example/docker-machine/`, `example/linux`, `example/github-auth`, etc examples which contain a `start.sh` which we can then use for local testing.

This was meant as a PR to bring the instructions up-to-date with what we are really doing, so let's get this merged first and then start a new issue/PR to discuss what we would like to do :grinning:

**Comment by betatim** [comment]
Merge?
